src/cls/queue/cls_queue_src.cc

```diff
@@ -0,0 +1,901 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "include/types.h"
+
+#include <errno.h>
+
+#include "objclass/objclass.h"
+#include "cls/queue/cls_queue_types.h"
+#include "cls/queue/cls_queue_ops.h"
+#include "cls/queue/cls_queue_const.h"
+#include "cls/queue/cls_queue.h"
+
+static int get_queue_head_and_size(cls_method_context_t hctx, cls_queue_head& head, uint64_t& head_size)
+{
+  //read head size
+  bufferlist bl_head_size;
+  int ret = cls_cxx_read(hctx, 0, sizeof(uint64_t), &bl_head_size);
+  if (ret < 0) {
+    CLS_LOG(0, "ERROR: get_queue_head_and_size: failed to read head with error %d\n", ret);
+    return ret;
+  }
+  //decode head size
+  auto iter = bl_head_size.cbegin();
+  try {
+    decode(head_size, iter);
+  } catch (buffer::error& err) {
+    CLS_LOG(0, "ERROR: get_queue_head_and_size: failed to decode head size \n");
+    return -EINVAL;
+  }
+
+  CLS_LOG(10, "INFO: get_queue_head_and_size: head size is %lu\n", head_size);
+
+  //read the head
+  bufferlist bl_head;
+  ret = cls_cxx_read2(hctx, sizeof(uint64_t), (head_size - sizeof(uint64_t)), &bl_head, CEPH_OSD_OP_FLAG_FADVISE_WILLNEED);
+  if (ret < 0) {
+    return ret;
+  }
+
+  //decode head
+  iter = bl_head.cbegin();
+  try {
+    decode(head, iter);
+  } catch (buffer::error& err) {
+    CLS_LOG(1, "ERROR: get_queue_head_and_size: failed to decode head\n");
+    return -EINVAL;
+  }
+
+  return 0;
+}
+
+int cls_create_queue(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+{
+  auto in_iter = in->cbegin();
+
+  cls_create_queue_op op;
+  try {
+    decode(op, in_iter);
+  } catch (buffer::error& err) {
+    CLS_LOG(1, "ERROR: cls_create_queue(): failed to decode entry\n");
+    return -EINVAL;
+  }
+
+  // create the object
+  int ret = cls_cxx_create(hctx, true);
+  if (ret < 0) {
+    CLS_LOG(0, "ERROR: %s(): cls_cxx_create returned %d", __func__, ret);
+    return ret;
+  }
+  
+  CLS_LOG(10, "INFO: cls_create_queue create queue of head size %lu", op.head_size);
+  CLS_LOG(10, "INFO: cls_create_queue create queue of size %lu", op.head.queue_size);
+
+  uint64_t head_size = QUEUE_HEAD_SIZE_1K;
+
+  if (op.head.has_urgent_data) {
+    if (op.head_size == 0) {
+      head_size = QUEUE_HEAD_SIZE_4K;
+      op.head.tail = op.head.front = QUEUE_START_OFFSET_4K;
+      op.head.last_entry_offset = QUEUE_START_OFFSET_4K;
+    } else {
+      head_size = op.head_size;
+      op.head.tail = op.head.front = head_size;
+      op.head.last_entry_offset = head_size;
+    }
+  } else {
+    head_size = QUEUE_HEAD_SIZE_1K;
```

Comment:
only the `has_urgent_data` case looks at `op.head_size`; does it ever make sense for the non-urgent-data case to customize the max head size?

Reply:
I think we had a discussion about having some extra space in the head if we may want to add fields in future, for that reason even the non urgent data case has a fixed head size.

Reply:
sorry, what i meant to say is that if a client creates a queue and passes `op.head_size=4k` and `op.head.has_urgent_data=false`, this else block ignores `op.head_size` and uses QUEUE_HEAD_SIZE_1K instead

Reply:
Matt wanted this config variable rgw_gc_queue_head_size to be added so that while testing, the optimum head size for GC could be arrived at and would be removed later, hence I haven't taken it into account in the non urgent data case.